dask/array/routines.py

```diff
@@ -630,10 +674,18 @@ def unique(ar, return_index=False, return_inverse=False, return_counts=False):
         out_dtype
     )
 
+    # Split out all results to return to the user.
+
     result = [out["values"]]
     if return_index:
         result.append(out["indices"])
     if return_inverse:
+        # Using the returned unique values and arange of unknown length, find
+        # each value matching a unique value and replace it with its
+        # corresponding index or `0`. There should be only one entry for this
+        # index in axis `1` (the one of unknown length). Reduce axis `1`
+        # through summing to get an array with known dimensionality and the
+        # mapping of the original values.
```

Comment:
The line below is potentially O(n**2), right?

Reply:
Where are you getting `O(n**2)` from? The comparison? Something else?

FWIW unique values are likely far fewer than values in the original array. So I doubt it is actually that bad in practice, but a no-op case could be a problem. Just bet it is a problem long before here. IOW memory usage in the reduction step.

Reply:
If `out` is O(n) where n is the size of `ar`, then `ar[:,None]` and `out[None,:]` will be O(n**2) in size (if only logical, not allocated), and therefore the comparison too, as well as the multiplication and sum that follows.

Reply:
So again sounds like you are referring to the comparison. This is a worst case estimate and is certainly true for that case. Namely the case when there are no unique values. In reality, I don't think someone will call `unique` without being reasonably sure it is worth the effort. So expect this will be much closer to `O(N)` in practice.

Though I think this will be the least of our concerns if we have no unique values on a large array. Namely the reduction step will be wielding some very large arrays and possibly run out of memory. The comparison here is at least chunked. The reduction step is really just concatenation of all of the chunks after running `unique` once. Honestly I don't know of an easy fix to this sort of problem. Maybe if one could keep track of the count of unique values only in a first pass to see if things will get unreasonable, but am not really sure how that would be possible.